### PR TITLE
Refactor `Pieces` class to `PiecesCollection`

### PR DESCRIPTION
**Pull request by jacobdekeizer**
Implement BaseResourceCollection so we can create more resource collections.

### CODE REVIEW

**Comment by Thorry84** [comment]
What responsibility does the BaseResourceCollection have? Why would Pieces simply be a collection? I would opt to not have a separate class until it is required and has added value.

**Comment by jacobdekeizer** [comment]
@Thorry84 The shipments api returns multiple shipments. If this pr gets merged, I will make a shipments collection so we can get multiple labels from the dhl api.

**Comment by Thorry84** [comment]
I understand, but why not make it a collection? What does the BaseResourceCollection class add, other than being an extra wrapper?

**Comment by jacobdekeizer** [comment]
@Thorry84 It adds a default constructor so you can add objects or arrays.
@mvdnbrk Do you have time to look at this PR?

**Comment by mvdnbrk** [comment]
At the moment I don't see the need for a `BaseResourceCollection`.

We could simply turn the `Pieces` class into a `PiecesCollection` class which extends the `Tightenco\Collect\SupportCollection` class.

If needed we can always refactor this to some sort of `BaseResourceCollection`.

File: src/Resources/Pieces.php [hunk 1]
```diff
@@ -2,11 +2,25 @@
 
 namespace Mvdnbrk\DhlParcel\Resources;
 
+use Mvdnbrk\DhlParcel\Contracts\Arrayable;
 use Mvdnbrk\DhlParcel\Exceptions\ResourceNotAccepted;
 use Tightenco\Collect\Support\Collection;
 
-class Pieces extends BaseResourceCollection
+class Pieces extends Collection implements Arrayable
 {
```
Comment (mvdnbrk) [inline src/Resources/Pieces.php]: `Tightenco\Collect\Support\Collection` already implements it's own `Arrayable` interface, so I think this can be removed.

Comment (mvdnbrk) [reply src/Resources/Pieces.php]: Please rename the class to `PiecesCollection`. To be clear this is a collection of items.

File: src/Resources/Pieces.php [hunk 2]
```diff
@@ -2,11 +2,25 @@
...
+    /**
+     * BaseResourceCollection constructor.
+     *
+     * @noinspection PhpMissingParentConstructorInspection
```
Comment (mvdnbrk) [inline src/Resources/Pieces.php]: Remove this line.

File: src/Exceptions/ResourceNotAccepted.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+     * Create a new Resource not accepted exception.
+     *
+     * @param  mixed  $resource
+     * @return static
+     */
+    public static function forResource($resource)
+    {
+        return new static('Resource for class ['.get_class($resource).'] not accepted.');
+    }
+}
```
Comment (mvdnbrk) [inline src/Exceptions/ResourceNotAccepted.php]: This Exception class is not used. Please remove it.

File: src/Resources/Pieces.php [hunk 4]
```diff
@@ -2,25 +2,22 @@
...
+use Mvdnbrk\DhlParcel\Contracts\Arrayable;
+use Mvdnbrk\DhlParcel\Exceptions\ResourceNotAccepted;
+use Tightenco\Collect\Support\Collection;
 
+class Pieces extends Collection implements Arrayable
+{
     /**
-     * Create a new Pieces resource.
+     * BaseResourceCollection constructor.
      *
```
Comment (mvdnbrk) [inline src/Resources/Pieces.php]: Change docblock: Create a new Pieces Collection instance.

File: src/Resources/Pieces.php [hunk 5]
```diff
@@ -2,25 +2,22 @@
...
-     * @param  array  $attributes
-     * @return void
+     * @noinspection PhpMissingParentConstructorInspection
+     * @param  array  $items
      */
-    public function __construct(array $attributes = [])
+    public function __construct($items = [])
     {
```
Comment (mvdnbrk) [inline src/Resources/Pieces.php]: Please be consistent. Change to `array $items = []`

File: src/Resources/Pieces.php [hunk 6]
```diff
@@ -30,29 +27,35 @@ public function __construct(array $attributes = [])
      * @param  \Mvdnbrk\DhlParcel\Resources\Piece|array  $value
      * @return void
      */
-    public function add($value)
+    public function add($value): void
     {
```
Comment (mvdnbrk) [inline src/Resources/Pieces.php]: Code consistency: remove `: void`

File: src/Resources/Pieces.php [hunk 7]
```diff
@@ -30,29 +27,35 @@ public function __construct(array $attributes = [])
...
     }
 
     /**
      * Convert the Pieces resource to an array.
      *
      * @return array
      */
-    public function toArray()
+    public function toArray(): array
     {
```
Comment (mvdnbrk) [inline src/Resources/Pieces.php]: Please remove `: array` for same reason as above. If you would like to change this do this in a separate PR.

File: src/Resources/Pieces.php [hunk 8]
```diff
@@ -30,29 +27,35 @@ public function __construct(array $attributes = [])
...
         }
 
-        $this->items[] = new Piece($value);
+        if (is_array($value)) {
+            $this->items[] = new Piece($value);
+
+            return;
+        }
+
+        throw ResourceNotAccepted::forResource($value);
```
Comment (mvdnbrk) [inline src/Resources/Pieces.php]: Do not throw an exception here.

Comment (jacobdekeizer) [reply src/Resources/Pieces.php]: I don't see a reason to not throw an exception here.
